### ISSUE TITLE
trouble with utf-8 cyrillic symbols

### ISSUE THREAD

**Issue by kir3d** [folz/cleverbot.py#21]
On website it works but received pattern |xxxx when tried:

from cleverbot import Cleverbot

cb = Cleverbot()
message=cb.ask('привет')

**Comment by basvdwollenberg** [comment]
Temporary work-around:
http://stackoverflow.com/a/11281948/5082094
```python
eval('u"""' + message.replace('|', r'\u').replace('"', r'\"') + '-"""')[:-1]
```

**Comment by kir3d** [comment]
Thank you very much!

**Comment by folz** [comment]
Strongly recommend you avoid "eval" because of security reasons related to eval-ing untrusted input.

That said, do either of you have suggestions for resolving this? Coming from english, I'm don't know what look for while working with non-ascii characters.

**Comment by kir3d** [comment]
I see threat 'eval'.

English work perfect, cyrillic answer likes '|041F|0440|0438|0432|0435|0442.'

Python3 prevents replace "|" to  "\u" (single backslash).

message.replace("|",r"\u") return '\\u041F\\u0440\\u0438\\u0432\\u0435\\u0442.' 
message.replace("|","\u") return SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 0-1: truncated \uXXXX escape

**Comment by basvdwollenberg** [comment]
> Strongly recommend you avoid "eval" because of security reasons related to eval-ing untrusted input.
You really expect Cleverbot to send back 'untrusted' input?

**Comment by folz** [comment]
"Untrusted" in this context just means "anything that wasn't generated by your program", so yes.

**Comment by basvdwollenberg** [comment]
Any way eval could be abused in this case?

**Comment by kir3d** [comment]
Cleverbot is self-learning? Can use users quotes? Yes. It's mean once time some user can wrote potential dangerous command and Cleverbot once time use this command like answer and eval with have enough permission can execute this command. IMHO.

**Comment by basvdwollenberg** [comment]
I just tested this:
```python
import os


print(eval('u"""' + "os.system('echo Test')".replace('|', r'\u').replace('"', r'\"') + '-"""')[:-1])
print(eval("os.system('echo Test')"))
```
Output:
```
os.system('echo Test')
Test
0
```
Not sure, but I believe that it is actually safe to use then?

**Comment by kir3d** [comment]
Try print(eval("os.system('clear')"))

**Comment by kir3d** [comment]
http://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html
http://stackoverflow.com/questions/13066594/is-there-a-way-to-secure-strings-for-pythons-eval
http://tav.espians.com/a-challenge-to-break-python-security.html

May be will help sandbox https://github.com/haypo/pysandbox or http://doc.pypy.org/en/latest/sandbox.html ?

**Comment by NyanKiyoshi** [comment]
Otherwise, you could extract the unicode data and replace it with the associated unicode character (by converting the UTF-16 data (hexadecimal) to a integer and then, chr on Python3 or unichr on Python2) by using a regex (or by using another way).

For example:
```python
>>> cleverbot_utf_16_word = re.compile(r'\|([0-9A-F]{4})')
>>> reponse = '|041F|0440|0438|0432|0435|0442.'
>>> re.sub(cleverbot_utf_16_word, lambda matchobj: chr(int(matchobj.group(1), 16)), response)
'Привет.'
```

It's pretty dirty but some improvements can be done :)

**Comment by kir3d** [comment]
NyanKiyoshi, nice! Thank you!

**Comment by basvdwollenberg** [comment]
Ah that's way better, thanks @NyanKiyoshi!

**Comment by SoupMeNot** [comment]
I have some trouble in using german unicode-characters, too. Unfortunately the method suggested by @NyanKiyoshi doesn't help. The problem shows up as following: I can send all unicode-characters without problem, just when the Bot is answering something like _ä,ö,ü,ß_ i get the following output:

> Traceback (most recent call last):
  File "cbtest.py", line 21, in <module>
    main()
  File "cbtest.py", line 16, in main
    print('>> Cleverbot: {}'.format(answer))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 1-2: ordinal not in range(128)

This problem is solved quite easily by leaving out the format()-function, i can convert those characters before output later easily.
The problem giving me more trouble is, that one message later the urlencode()-function is giving me the following traceback:

> >> You: Tschau!
>> Cleverbot: {}TschÃ¼ss.
>> You: Tschau Tschau
Traceback (most recent call last):
  File "cbtest.py", line 21, in <module>
    main()
  File "cbtest.py", line 15, in main
    answer = cleverbot_client.ask(question)
  File "/home/user/dev/python/cleverbot.py", line 96, in ask
    resp = self._send()
  File "/home/user/dev/python/cleverbot.py", line 133, in _send
    enc_data = urlencode(self.data)
  File "/usr/lib/python2.7/urllib.py", line 1349, in urlencode
    v = quote_plus(str(v))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 4-5: ordinal not in range(128)

(I already feed the _parse()-function with the RE-patch from @NyanKiyoshi, so the input to urlencode should be already cleaned up)
When i try a workaround like http://stackoverflow.com/questions/6480723/, converting erverything properly to the crappy iso-8859-1 (just to be sure to use the same encoding as the cb-page, as utf-8 has some little differences), i get an DENIED error from the CB-page. Probably the session-token is calculated wrong. But at least the urlencode-function() is working and i get an answer from the server.

Some good questions to test this is to bother the guy with questions like "Hallo" (hello) until it is speaking in german to you and "Tschüss!" (good bye), probably it will answer to "Tschüss" in the same way and you get your testing-output with unicode-char from the bot. Otherwise take some single sentences from german/<other unicode languages> - wikipedia sites to feed it.

**Comment by NyanKiyoshi** [comment]
Does it works like this? https://github.com/NyanKiyoshi/cleverbot.py/commit/4a25053f0bc295c0b4968f256202b280afe38003

```
  [02:43:40]  @Chocolat | !q Du bist eine mädchen?
  [02:43:41]  @WuWu     | Chocolat: ja ich bin ein Mädchen und 14 Jahre alt.
```

**Comment by Y4kuzi** [comment]
I just tested above example, and I got this as reply:

```
>>> cb.ask('Du bist eine mädchen?')
'Was bist du Junge oder MÃ¤dchen?'
```

**Comment by mrgigabyte** [comment]
Y4kuzi, check the pull request i have fixed that issue :)

**Comment by folz** [comment]
For those following along at home, please update to v1.1.0 on pypi and let me know if this fixes your problem. It includes @mrgigabyte's decode-encode fix from #33.
